src/visualization/plot_recipes.jl

```diff
@@ -288,10 +288,55 @@ end
                    solution_variables=cons2prim,
                    grid_lines=true, max_supported_level=11, nvisnodes=nothing, slice_axis=:z,
                    slice_axis_intercept=0)
-  return PlotData2D(sol;
-                    solution_variables=solution_variables,
-                    grid_lines=grid_lines, max_supported_level=max_supported_level,
-                    nvisnodes=nvisnodes, slice_axis=slice_axis,
-                    slice_axis_intercept=slice_axis_intercept)
+
+  mesh, _, _, _ = mesh_equations_solver_cache(sol.prob.p)
+  if ndims(mesh) == 1
+    return PlotData1D(sol;
+                      solution_variables=solution_variables,
+                      grid_lines=grid_lines, max_supported_level=max_supported_level,
+                      nvisnodes=nvisnodes, slice_axis=slice_axis,
+                      slice_axis_intercept=slice_axis_intercept)
+  else
+    return PlotData2D(sol;
+                      solution_variables=solution_variables,
+                      grid_lines=grid_lines, max_supported_level=max_supported_level,
+                      nvisnodes=nvisnodes, slice_axis=slice_axis,
+                      slice_axis_intercept=slice_axis_intercept)
+    end
```

Comment:
Did you check that this is actually type-unstable (using `@code_warntype` or something similar)? As far as I know, the compiler might be able to figure out which path to take since `ndims(mesh)` depends on the type of `mesh`, which is part of the type of `sol.prob.p`, which is part of the type of `sol`. If the compiler decides that this is too complicated, it's probably not worth optimizing this here, since this part is definitely not performance critical and uses a function barrier internally.

Reply:
You seem to be right: I copied the relevant lines from the recipe to a new (but identical) function `make_pd` and got this:
```julia
julia> @code_warntype Trixi.make_pd(sol) # sol is from a 1D simulation
Variables
  #self#::Core.Compiler.Const(Trixi.make_pd, false)
  sol::ODESolution{Float64,2,Array{Array{Float64,1},1},Nothing,Nothing,Array{Float64,1},Array{Array{Array{Float64,1},1},1},ODEProblem{Array{Float64,1},Tuple{Float64,Float64},true,SemidiscretizationHyperbolic{TreeMesh{1,Trixi.SerialTree{1}},LinearScalarAdvectionEquation1D{Float64},typeof(initial_condition_convergence_test),Trixi.BoundaryConditionPeriodic,Nothing,DG{Float64,LobattoLegendreBasis{Float64,4,Array{Float64,2},StaticArrays.SArray{Tuple{4,2},Float64,2,8},StaticArrays.SArray{Tuple{4,4},Float64,2,16}},Trixi.LobattoLegendreMortarL2{Float64,4,StaticArrays.SArray{Tuple{4,4},Float64,2,16}},typeof(flux_lax_friedrichs),VolumeIntegralWeakForm},NamedTuple{(:elements, :interfaces, :boundaries),Tuple{Trixi.ElementContainer1D{Float64,1,3},Trixi.InterfaceContainer1D{Float64,1,3},Trixi.BoundaryContainer1D{Float64,1,3}}}},ODEFunction{true,typeof(Trixi.rhs!),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,typeof(SciMLBase.DEFAULT_OBSERVED),Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},SciMLBase.StandardODEProblem},CarpenterKennedy2N54{typeof(OrdinaryDiffEq.trivial_limiter!),typeof(OrdinaryDiffEq.trivial_limiter!)},OrdinaryDiffEq.InterpolationData{ODEFunction{true,typeof(Trixi.rhs!),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,typeof(SciMLBase.DEFAULT_OBSERVED),Nothing},Array{Array{Float64,1},1},Array{Float64,1},Array{Array{Array{Float64,1},1},1},OrdinaryDiffEq.LowStorageRK2NCache{Array{Float64,1},Array{Float64,1},OrdinaryDiffEq.LowStorageRK2NConstantCache{4,Float64,Float64},typeof(OrdinaryDiffEq.trivial_limiter!),typeof(OrdinaryDiffEq.trivial_limiter!)}},DiffEqBase.DEStats}

Body::PlotData1D{Array{Float64,1},Array{Float64,1},StaticArrays.SArray{Tuple{1},String,1,1},Array{Float64,1}}
1 ─ %1 = Trixi.:(var"#make_pd#477")(Trixi.cons2prim, true, 11, Trixi.nothing, :z, 0, #self#, sol)::PlotData1D{Array{Float64,1},Array{Float64,1},StaticArrays.SArray{Tuple{1},String,1,1},Array{Float64,1}}
└──      return %1

julia> @code_warntype Trixi.make_pd(sol) # sol is from a 2D simulation
Variables
  #self#::Core.Compiler.Const(Trixi.make_pd, false)
  sol::ODESolution{Float64,2,Array{Array{Float64,1},1},Nothing,Nothing,Array{Float64,1},Array{Array{Array{Float64,1},1},1},ODEProblem{Array{Float64,1},Tuple{Float64,Float64},true,SemidiscretizationHyperbolic{TreeMesh{2,Trixi.SerialTree{2}},LinearScalarAdvectionEquation2D{Float64},typeof(initial_condition_convergence_test),Trixi.BoundaryConditionPeriodic,Nothing,DG{Float64,LobattoLegendreBasis{Float64,4,Array{Float64,2},StaticArrays.SArray{Tuple{4,2},Float64,2,8},StaticArrays.SArray{Tuple{4,4},Float64,2,16}},Trixi.LobattoLegendreMortarL2{Float64,4,StaticArrays.SArray{Tuple{4,4},Float64,2,16}},typeof(flux_lax_friedrichs),VolumeIntegralWeakForm},NamedTuple{(:elements, :interfaces, :boundaries, :mortars, :fstar_upper_threaded, :fstar_lower_threaded),Tuple{Trixi.ElementContainer2D{Float64,1,3},Trixi.InterfaceContainer2D{Float64,1,3},Trixi.BoundaryContainer2D{Float64,1,3},Trixi.L2MortarContainer2D{Float64,1,3},Array{StaticArrays.MArray{Tuple{1,4},Float64,2,4},1},Array{StaticArrays.MArray{Tuple{1,4},Float64,2,4},1}}}},ODEFunction{true,typeof(Trixi.rhs!),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,typeof(SciMLBase.DEFAULT_OBSERVED),Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},SciMLBase.StandardODEProblem},CarpenterKennedy2N54{typeof(OrdinaryDiffEq.trivial_limiter!),typeof(OrdinaryDiffEq.trivial_limiter!)},OrdinaryDiffEq.InterpolationData{ODEFunction{true,typeof(Trixi.rhs!),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,typeof(SciMLBase.DEFAULT_OBSERVED),Nothing},Array{Array{Float64,1},1},Array{Float64,1},Array{Array{Array{Float64,1},1},1},OrdinaryDiffEq.LowStorageRK2NCache{Array{Float64,1},Array{Float64,1},OrdinaryDiffEq.LowStorageRK2NConstantCache{4,Float64,Float64},typeof(OrdinaryDiffEq.trivial_limiter!),typeof(OrdinaryDiffEq.trivial_limiter!)}},DiffEqBase.DEStats}

Body::PlotData2D{Array{Float64,1},Array{Array{Float64,2},1},StaticArrays.SArray{Tuple{1},String,1,1},Array{Float64,2}}
1 ─ %1 = Trixi.:(var"#make_pd#477")(Trixi.cons2prim, true, 11, Trixi.nothing, :z, 0, #self#, sol)::PlotData2D{Array{Float64,1},Array{Array{Float64,2},1},StaticArrays.SArray{Tuple{1},String,1,1},Array{Float64,2}}
└──      return %1
```

At least to me it looks like it is type stable after all.